twisted/test/test_twistd.py

```diff
@@ -1802,7 +1857,7 @@ def startService(self):
         self.assertEqual(
             self.mockos.actions,
             [('chdir', '.'), ('umask', 0o077), ('fork', True), 'setsid',
-             ('fork', True), ('write', -2, '1 ' + 'x' * 98),
+             ('fork', True), ('write', -2, b'1 ' + b'x' * 98),
```

Comment:
I like the sound of this a lot, but what would it test?  The code should never receive partial utf-8 sequences.  Here are the cases in which Unicode can be written to the pipe:

1. It's Python 2 and the exception raised in the child has a `unicode` message.  This can happen in lots of ways (`<some bytes> % (<some unicode>,)` -> `<unicode>`!)   In this case `traceback.format_exception_only` encodes the object into 'ascii' with 'backslashreplace', so we might see partial backslash representations (e.g. `"u'\N{BULLET}'"` -> '\u20`) if the code point occurs around the 100 byte mark.
1. It's Python 3.  In this case the exception message is truncated to 100 code points before encoding, so we might have *more* than 100 bytes, but the backslash representations will be complete.

In other words, only in the Python 2 case will truncation occur after encoding, but it will be on backslash representations of code points, not their UTF-8 encoding.

Since the read side of the pipe, which this test exercises, is kind of a private API, or at least can't exist without the write side I just described, I don't think it needs to be tested against partial UTF-8 sequences.  I do think I'll test it against partial backslash representations, though, as that truncation can definitely occur.

Does that sound reasonable?